### PR TITLE
PITR (draft PR)

### PR DESCRIPTION
**Pull request by arindamnayak**

### CODE REVIEW

File: go/mysql/flavor_mysql.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ type mysqlFlavor struct{}
 
 // masterGTIDSet is part of the Flavor interface.
 func (mysqlFlavor) masterGTIDSet(c *Conn) (GTIDSet, error) {
-	qr, err := c.ExecuteFetch("SELECT @@GLOBAL.gtid_executed", 1, false)
+	qr, err := c.ExecuteFetch("SELECT @@global.gtid_executed", 1, false)
```
Comment (deepthi) [inline go/mysql/flavor_mysql.go:36]: any particular reason you needed to change this?

File: go/vt/vttablet/tabletmanager/restore.go [hunk 2]
```diff
@@ -44,6 +49,13 @@ var (
 	restoreFromBackup     = flag.Bool("restore_from_backup", false, "(init restore parameter) will check BackupStorage for a recent backup at startup and start there")
 	restoreConcurrency    = flag.Int("restore_concurrency", 4, "(init restore parameter) how many concurrent files to restore at once")
 	waitForBackupInterval = flag.Duration("wait_for_backup_interval", 0, "(init restore parameter) if this is greater than 0, instead of starting up empty when no backups are found, keep checking at this interval for a backup to appear")
+
+	// Flags for PITR
+	restoreToTimeStr = flag.String("restore_to_time", "", "(init restore parameter) will restore to the specified time, it depends on the binlog related flags.")
```
Comment (deepthi) [inline go/vt/vttablet/tabletmanager/restore.go]: are we requiring that the timestamp set in the recovery keyspace and this time should match?
what happens if someone provides this though they are not actually doing a PITR?

File: go/vt/vttablet/tabletmanager/restore.go [hunk 3]
```diff
@@ -166,6 +185,88 @@ func (agent *ActionAgent) restoreDataLocked(ctx context.Context, logger logutil.
...
+				break
+			}
+		}
+		return nil
+	})
+	return gtid
+}
+
+// replicateUptoGTID replicates upto specified gtid from binlog server
+func (agent *ActionAgent) replicateUptoGTID(ctx context.Context, gtid string) error {
```
Comment (deepthi) [inline go/vt/vttablet/tabletmanager/restore.go]: suggestion: catchupToGTID

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go [hunk 4]
```diff
@@ -169,8 +169,10 @@ func (vs *vstreamer) Stream() error {
 
 	// Ensure sh is Open. If vttablet came up in a non_serving role,
 	// the schema engine may not have been initialized.
-	if err := vs.sh.Open(); err != nil {
-		return wrapError(err, vs.pos)
+	if !vs.se.SkipMetaCheck {
```
Comment (deepthi) [inline go/vt/vttablet/tabletserver/vstreamer/vstreamer.go]: why did we need this?

File: go/vt/dbconfigs/dbconfigs.go [hunk 5]
```diff
@@ -394,6 +394,12 @@ func (dbcfgs *DBConfigs) getParams(userKey string, dbc *DBConfigs) (*UserConfig,
 	return uc, cp
 }
 
+// SetDbParams sets the dba and app params
+func (dbcfgs *DBConfigs) SetDbParams(params mysql.ConnParams) {
```
Comment (deepthi) [inline go/vt/dbconfigs/dbconfigs.go]: this feels too specific to the usage. If it is going in here, maybe it should be
```
SetDbParams(dbaParams, appParams mysql.ConnParams)
```

File: go/mysql/flavor_mysql.go (hunk 1 above)
Comment (arindamnayak) [reply go/mysql/flavor_mysql.go:36]: ripple does not like `GLOBAL` , I think we should add some failsafe way to fix the case sensitiveness.

File: go/vt/vttablet/tabletmanager/restore.go (hunk 2 above)
Comment (arindamnayak) [reply go/vt/vttablet/tabletmanager/restore.go]: ohh I forgot, I though snapshot time = backup time, I will change my code.

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go (hunk 4 above)
Comment (arindamnayak) [reply go/vt/vttablet/tabletserver/vstreamer/vstreamer.go]: for getting GTID from timestamp, we don't need the query to the information_schema tablet meta information fetch and with ripple, this query is not supported

File: go/vt/dbconfigs/dbconfigs.go (hunk 5 above)
Comment (arindamnayak) [reply go/vt/dbconfigs/dbconfigs.go]: I did not find the method to set those params and those params are being used by the engine internally to connect the binlog server, that is why I added this.

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go (hunk 4 above)
Comment (deepthi) [reply go/vt/vttablet/tabletserver/vstreamer/vstreamer.go]: ah ok.

File: go/vt/vttablet/tabletmanager/restore.go [hunk 6]
```diff
@@ -166,6 +187,105 @@ func (agent *ActionAgent) restoreDataLocked(ctx context.Context, logger logutil.
...
+		}
+	}
+	return ""
+}
+
+// replicateUptoGTID replicates upto specified gtid from binlog server
+func (agent *ActionAgent) catchupToGTID(ctx context.Context, gtid string) error {
+	gtid = strings.Replace(gtid, "MySQL56/", "", 1)
+
+	gtidNew := strings.Split(gtid, ":")[0] + ":" + strings.Split(strings.Split(gtid, ":")[1], "-")[1]
```
Comment (deepthi) [inline go/vt/vttablet/tabletmanager/restore.go]: take a look at https://github.com/vitessio/vitess/blob/master/go/mysql/mysql56_gtid_set.go. maybe you don't need to do all this manual processing of the GTID string.

Comment (arindamnayak) [reply go/vt/vttablet/tabletmanager/restore.go]: yes, I was planning to do it, this is a quick hack to get started and make it working.

**Pull request merged**